### ISSUE TITLE
Cross-Compatibility of Python 2 and 3 in a single folder.

### ISSUE THREAD

**Issue by slee047** [agronholm/pythonfutures#69]
Hi, I have a corporate project that requires maintain of compatibility between 2.7 and 3.X (Mainly 3.6.Y).

Due to Firewall policy, I could not use pip install X for installing various dependencies that requires futures. This includes _apscheduler_ etc.

The only way for me to install these modules is via `--install-lib` (More on:
 [https://docs.python.org/2/install/index.html](url)) option with the use of environment variable `PythonPath` with the use of .tar / .zip version.

So, I have installed both Python 2.7 and 3.6 on separate folders in _C:\_ and a common Module Path at _C:\Python_Modules_. This will result _apscheduler_ crash for Python 3 version. This is because _apscheduler_ for Python 3 tries to import futures from Python 2 egg regardless if I tried to compile Futures for Python 3 or not.

Attempted to compile Futures in Python 3 resulted a redundancy warning.
![redundant for python3](https://user-images.githubusercontent.com/25880359/32875780-a851b90a-cad4-11e7-8fdc-b0046c3cd970.png)

The first one is:
`raise exception_type, self._exception, self._traceback` in line 381 of _base.py
![first issue](https://user-images.githubusercontent.com/25880359/32875527-67d940e2-cad3-11e7-961c-4a9a387992c5.png)

The fix is to change the file (futures-3.1.1\build\lib\concurrent\futures\_base.py) with these:
1. Change the line 381 code to:
`raise_with_traceback(exception_type(self._exception))`
2: Add this line of code below line 9 (`import types`) so as to import the traceback feature.
`from future.utils import raise_with_traceback`

The changed code (Change the extension from txt to py) can be downloaded here:
[_base.txt](https://github.com/agronholm/pythonfutures/files/1477434/_base.txt)

The second one is the issue of the Queue and queue for Python 2 and 3 compatibility.
![second issue](https://user-images.githubusercontent.com/25880359/32876004-afbc334a-cad5-11e7-9099-ea291f8277a3.png)

This is also faced by other modules such as [https://github.com/docker/compose/issues/2108](url)

The fix is to change the file (futures-3.1.1\build\lib\concurrent\futures\thread.py) with this on line 8:
```
try:
	import Queue as queue
except ImportError:
	import queue
```

The changed code (Change the extension from txt to py) can be downloaded here:
[thread.txt](https://github.com/agronholm/pythonfutures/files/1477458/thread.txt)

With these changes on both files, I will able to solve this issue of futures module not able to use in Python 3.

**Comment by agronholm** [comment]
Please do not attempt to install futures on Python 3. It's already part of the standard library there. For this reason I will never add any py3 compatibility code to it.

**Comment by slee047** [comment]
No, the issue is **NOT about attempting to install futures on Python 3**.

The issue is about other modules who has _futures_ as dependency will _always_ use the Python 2 version when these dependencies are installed in a common folder. (Instead of using pip)

This usually happens when user uses this command to install their dependencies / other modules:
`C:\Python<Version>\python.exe setup.py install --install-lib <Path in environment variable 'PythonPath'>`

Just like the first IDLE screenshot, I have compiled a APScheduler-3.4.0-py3.6.egg, but it still tried to use `futures` from Python 2.

**Comment by agronholm** [comment]
Well, it's certainly not a problem with apscheduler attempting to *install* futures (there's a conditional dependency). So you're saying that you've hacked the python path in a way that will override standard library imports with something from an external directory?

**Comment by slee047** [comment]
Hi, I  don't think I am hacking when I am following one of the installation instructions in official Python webpage here:
[https://docs.python.org/2/install/index.html](https://docs.python.org/2/install/index.html)

As stated previously, I cannot do the modern pip install due to corporate Internet policy that blocks the download.

Thus, I need to install these modules manually. One way is to install in default location, but then managing the updates of these modules would be very tedious as I need to monitor 2 folders.

**Comment by agronholm** [comment]
> As stated previously, I cannot do the modern pip install due to corporate Internet policy that blocks the download.

You need to transfer the distributions there somehow, yes? Then why not transfer the wheels and then use pip to install them locally?

> Thus, I need to install these modules manually. One way is to install in default location, but then managing the updates of these modules would be very tedious as I need to monitor 2 folders.

This is, however, the norm due to the fact that Python 2 and 3 binary extensions are mutually incompatible.

> However, when I install both Python 2 and Python 3 version of it in the same folder, the Python 3 APScheduler will automatically tries to import futures, which I only installed using Python 2.

If you insist on the one directory approach then I have a better alternative. Setting `PYTHONPATH` prepends the path to `sys.path`, so what you want is to *append* to it. To do this, create a `sitecustomize.py` module somewhere in the regular Python path, and add the code that appends your custom package directory to `sys.path`. This will prevent Python from loading the wrong `concurrent.futures` module.

**Comment by slee047** [comment]
Thanks for the suggestion.

After considering the pros and cons of each option, I think it might be better to go _without_ `--install-lib` option. Therefore, `PythonPath` environment variable is no longer required in this instance.

This is to ensure each Python Modules can import the correct version of modules without trying to customize Python beyond the installer interface given that I need to think about the app's maintenance (Updates) and recovery (Re-installation from a new / re-formatted PC) steps in a holistic manner.

Before I came to this conclusion, I had tried to add multiple paths for `PythonPath`. I.e.: C:\Python_Modules\Py2 and C:\Python_Modules\Py3. 

Unfortunately this idea didn't work at all, so I have to abandon this idea of using single folder to host binaries for multiple Python versions. It is somehow a bit sad because this compilation from .tar.gz usually produce either a folder name or file with the format: _<module>-py2.7.egg_ / _<module>-py3.6.egg_ To be frank, the only price is that I cant see what are modules installed directly but one can always find the installation directory by other methods (Default: _C:\Python\Lib\site-packages_).

Thanks for the help and suggestion though :).
